hazelcast/src/test/java/com/hazelcast/cache/CacheTypesConfigTest.java

```diff
@@ -152,10 +154,19 @@ public void cacheConfigShouldBeAddedOnJoiningMember_whenCacheLoaderFactoryNotRes
         assertClusterSize(2, hz1, hz2);
 
         ICache<String, Person> cache = hz2.getCacheManager().getCache(cacheName);
-        String key = generateKeyOwnedBy(hz2);
-        expect.expectCause(new RootCauseMatcher(ClassNotFoundException.class, "classloading.domain.PersonCacheLoaderFactory - "
-                + "Package excluded explicitly"));
-        cache.invoke(key, new PersonEntryProcessor());
+        ClassLoader oldClassLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            Thread.currentThread().setContextClassLoader(hz2.getConfig().getClassLoader());
+            CacheProxy cacheProxy = (CacheProxy) cache;
+            CacheService cacheService = (CacheService) cacheProxy.getService();
+            expect.expectCause(new RootCauseMatcher(ClassNotFoundException.class, "classloading.domain.PersonCacheLoaderFactory - "
+                    + "Package excluded explicitly"));
+            cacheService.getCacheConfig(cache.getPrefixedName()).getCacheLoaderFactory();
```

Comment:
Why was this test changed? The second `HazelcastInstance` is anyway started with the filtering class loader and `invoke` will trigger the cache loader factory instantiation attempt. In the new version of the code, the following 2 lines will be anyway skipped because this line will throw the `ClassNotFoundException`.

Reply:
Because this test always succeeded, when it should have failed. With this change, class loading is more realistic and it makes for a better test.
I changed it because it was succeeding when I introduced it into app server (actual class loading environment) but it should have failed.

Reply:
> Because this test always succeeded, when it should have failed. With this change, class loading is more realistic and it makes for a better test.

The original test expects the exception to be thrown from Hazelcast partition operation threads. These threads are the ones which execute the `EntryProcessor` from `invoke` and are responsible for instantiating and invoking the `CacheLoader`. They use the config classloader (that is the `FilteringClassLoader` setup in second instance's `Config`).

The updated code changes the test-executing thread's context classloader and forces initialization of the `CacheLoader` from the test thread, triggering the exception to be thrown earlier from `getCacheLoaderFactory` call and does not exercise the partition operation thread code that should trigger the `CacheLoader` execution at all.

Can you elaborate on the setup that succeeded when it should have failed?

Reply:
> Can you elaborate on the setup that succeeded when it should have failed?

This was 3 years ago, I don't remember. All I know this was failing in my tests an I fixed it.

I reverted the changes and added another test, hope that resolves this (long) discussion